### ISSUE TITLE
Better error messages

### ISSUE THREAD

**Issue by msberends** [moodymudskipper/typed#4]
Thank you for this great package!

I was thinking about the language used in errors thrown by this package. They should ideally be perfectly understandable by users, and I would think the word ‘prototype’ is probably not really understood well by most (non-developer) users as it is a bit un-R-y.  

```r 
#> Error: `x`'s prototype should be numeric(0)
```

Also, the word ‘numeric’ could just be a word on is own, and does not need the definition of it being of length 0. It makes the errors unnecessarily complicated. Could you probably just throw;

```r 
#> Error: `x` must be of type numeric
```

?

Would be great to add your package to our package! But being our users’ advocate, I would like a bit clearer understandability :)

**Comment by moodymudskipper** [comment]
Thanks. I think you are right, these should be clearer. This "numeric(0)", comes from `deparse1(vctrs::vec_ptype(numeric()))`. This will be related to #3.

**Comment by moodymudskipper** [comment]
Hi @msberends ,

Can you try to install the branch "iteration2" with `remotes::install_github("moodymudskipper/typed@iteration2")`.

It will now work like this, which I think you will prefer.

![image](https://user-images.githubusercontent.com/18351714/96672740-395db800-1365-11eb-935a-17ef37ca300c.png)

If you can give me feedback on the new interface it will be extremely helpful as I think I'll wait a bit before advertising it to a wider audience.

**Comment by moodymudskipper** [comment]
We can still do better though. The following additional features would be nice :

* print the call that failed : we wight need some tryCatch magic in the active binding function, so we don't need to complexity errors in assertions.
* assertions from other packages, which might not use `call. = TRUE`, such as {assertive} functions, print non useful info : I think it could be handled by the above.
* error does not say if it happens due to return type, argument type, or variable type. If the above is well done the issue is mitigated. But the holy grail would be to be explicit about it. Arguments are doable I think, because `declare` knows it's being fed a promise. return type is harder, maybe we can make sure the last call is wrapped in return(), and have our assertions look up the stack, but it won't work with assertions not built with new_type_checker. We could override return, it would only affect functions of packages that import typed, but users won't like it. We could define our own returning function and have `typed_return(value, Integer())` but we'll need rlang or copy return_from from there, and it won't be highlighted like return. Or, maybe the best for the end, we do some more magic in `?` so we can type `Integer() ? return(value)`, we'll need return_from too.
* formulas passed to type checkers should use waldo before stopping with given error, and we should have a default error message and make the lhs optional.

**Comment by moodymudskipper** [comment]
A lot of improvements were done, error messages do their best to say where they happen, and they're very explicit, using : 


``` r
library(typed, warn.conflicts = FALSE)

##~~~~~~~~~~~~~~~~~~  wrong argument to function

test <- ? function(x= ? Integer()) {
  x
}

test(1)
#> Error: In `test(1)` at `Integer() ? x`: wrong argument to function, type mismatch
#> `typeof(value)`: "double" 
#> `expected`:      "integer"

##~~~~~~~~~~~~~~~~~~  wrong variable assignment

test <- function() {
  Integer() ? x
  x <- 1
}

test()
#> Error: In `test()` at `x <- ...`: type mismatch
#> `typeof(value)`: "double" 
#> `expected`:      "integer"

#~~~~~~~~~~~~~~~~~~ wrong return value

test <- Integer() ? function() {
  1
}

test()
#> Error: In `test()` at `Integer() ? return(1)`: wrong return value, type mismatch
#> `typeof(value)`: "double" 
#> `expected`:      "integer"
```

They will change again slightly, for the better, in iteration3.

closed by f42f7b5bdff69e80e9ed904b8859a9e473f7991c

**Comment by msberends** [comment]
What a great improvement you’ve made! I think this is indeed much more clear.

Great work!
